### PR TITLE
Add PoseUploader to stream poses to a PoseService

### PR DESCRIPTION
**Pull request by cschuet**

### CODE REVIEW

**Comment by ojura** [comment]
Could throw in a description what this is for? Sounds interesting.

File: cartographer/cloud/internal/map_builder_server.cc [hunk 1]
```diff
@@ -51,8 +51,11 @@ const common::Duration kPopTimeout = common::FromMilliseconds(100);
 
 MapBuilderServer::MapBuilderServer(
     const proto::MapBuilderServerOptions& map_builder_server_options,
-    std::unique_ptr<mapping::MapBuilderInterface> map_builder)
-    : map_builder_(std::move(map_builder)) {
+    MapBuilderServer::MapBuilderFactory map_builder_factory)
```
Comment (gaschler) [inline cartographer/cloud/internal/map_builder_server.cc]: This pattern is also necessary for cartographer_ros Node and MapBuilderBridge. Is there an alternative to specify the global SLAM callback later?

File: cartographer/cloud/internal/map_builder_server.h [hunk 2]
```diff
@@ -69,12 +71,16 @@ class MapBuilderContext : public MapBuilderContextInterface {
...
+      std::function<std::unique_ptr<mapping::MapBuilderInterface>(
+          const mapping::proto::MapBuilderOptions&,
+          mapping::PoseGraph::GlobalSlamOptimizationCallback)>;
   friend MapBuilderContext<mapping::Submap2D>;
   friend MapBuilderContext<mapping::Submap3D>;
 
   MapBuilderServer(
       const proto::MapBuilderServerOptions& map_builder_server_options,
-      std::unique_ptr<mapping::MapBuilderInterface> map_builder);
+      MapBuilderFactory map_builder_factory);
```
Comment (gaschler) [inline cartographer/cloud/internal/map_builder_server.h]: This change should be made in a separate commit with a follow up in cartographer_ros

File: cartographer/cloud/internal/pose_uploader.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+#include "cartographer/common/optional.h"
+#include "cartographer/mapping/id.h"
+#include "cartographer/transform/rigid_transform.h"
+
+namespace cartographer {
+namespace cloud {
+
+class PoseUploaderInterface {
+ public:
+  virtual void EnqueueLocalPose(common::Time time, int trajectory_id,
```
Comment (gaschler) [inline cartographer/cloud/internal/pose_uploader.h]: User shouldn't need to call these, only *active_trajectory should be public

File: cartographer/cloud/internal/pose_uploader.cc [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                                     const transform::Rigid3d &local_pose,
+                                     const transform::Rigid3d &global_pose) {
+  CHECK_EQ(node_id.trajectory_id, active_trajectory_id_);
+  last_global_node_id_.node_index = node_id.node_index;
+  // TODO(cschuet): Enqueue pose for uploading.
+}
+
+void PoseUploader::SetActiveTrajectory(int trajectory_id) {
+  active_trajectory_id_ = trajectory_id;
+  last_global_node_id_ = mapping::NodeId{trajectory_id, -1};
```
Comment (gaschler) [inline cartographer/cloud/internal/pose_uploader.cc]: use optional instead of -1

**Comment by cschuet** [comment]
@gaschler Thanks for taking a look. My thinking around this feature actually changed so I will go ahead and close this pull request. I don't think this code should be living as part of Cartographer, rather it belongs into a separate binary. I will send a PR to add a gRPC method to `MapBuilderService` to subscribe to global SLAM updates (similar to what we have for local SLAM updates).

@ojura I was working on the possibility to hook up Cartographer to a system that keeps track of robotic poses.

**Pull request closed without merging**